### ISSUE TITLE
artifacts with nvidia-340.xxx legacy driver

### ISSUE THREAD

**Issue by NiceandGently** [noodlylight/fusilli#17]
In mate it is possible to start the session directly with a windows-manager of your choice.
[rave@mother ~]$ gsettings get  org.mate.session.required-components windowmanager
'fusilli-mate-gtk'
```
[rave@mother ~]$ cat /usr/bin/fusilli-mate-gtk 
#!/bin/bash

function runFusilli() {
	fusilli --replace $@ 
}

ISSW=`glxinfo | grep "Software Rasterizer" -c`

# Try with direct rendering
HAVETFP=`glxinfo | grep texture_from_pixmap -c`

if ( [ $ISSW == 0 ] && [ $HAVETFP -gt 2 ] ); then 
	runFusilli $@
fi

# Try again with indirect rendering
export LIBGL_ALWAYS_INDIRECT=1

HAVETFP=`glxinfo | grep texture_from_pixmap -c`

if ( [ $ISSW == 0 ] && [ $HAVETFP -gt 2 ] ); then 
	runFusilli $@
fi

# Fall back to marco
exec marco $@
```
This script is based on a sript for compiz-0.88 which i use for fedora.
http://pkgs.fedoraproject.org/cgit/compiz.git/tree/compiz-mate-gtk
The fusilli script works well if i use it for a desktop starter inside a running session, but not if i set it in dconf for mate-session. Fussili starts without a windows-decorator :/
Do i miss something in the script?
Maybe the prob can be solved if you re-add the --marco option to gtk-windows-decorator.
This makes it possible to ask dconf for current selected metacity theme in mate session.
This is the base condition of my compiz-mate-gtk script to start compiz directly in session.

**Comment by NiceandGently** [comment]
Ok, i think the prob is that dconf isn't ready if i start fusilli directly with session, or there is a prob with reading the user gsetting keys in this early state of session.

Btw. after i updated fusilli to git version from 2015-03-08 i've serious probs with typing something with the keyboard, there is delay, which makes it hard for me to use fusilli.
Do you noticed something like this?

**Comment by noodlylight** [comment]
> Btw. after i updated fusilli to git version from 2015-03-08 i've serious probs with typing something with the keyboard, there is delay, which makes it hard for me to use fusilli.
> Do you noticed something like this?

No. fusilli works ok here. There were no significant changes during the last month. Do you remember which was the last version that worked ok for you?

> The fusilli script works well if i use it for a desktop starter inside a running session, but not if i set it in dconf for mate-session. Fussili starts without a windows-decorator :/

Hmm. Try using an absolute path in fsm->decoration->Command (i.e. /usr/bin/fusilli-decorator-gtk --replace)

> Maybe the prob can be solved if you re-add the --marco option to gtk-windows-decorator.

Hmm. I can probably do that, if it's necessary.

**Comment by NiceandGently** [comment]
I'm not exactly shure, but i think older git version was from the begining or middle january.
I will test the absolute path possibility in fussili config.

**Comment by NiceandGently** [comment]
Uhh, i fogot to compile fusilli with prefix=/usr, now fusilli runs a bit better, but ie. it was imposible to write this text if fusilli is running. I'm using firefox, but in a terminal or xchat typing was OK now.
Using /usr/bin/fusilli-decorator-gtk --replace in fsm doesn't help, also adding a sleep with 5 sec to my script didn't help to start fusilli with decorator here.
Weird that my script works well if call it with a desktop starter.

**Comment by NiceandGently** [comment]
Ok, i found a hint in .x-session-errors .
```
mate-session[27944]: WARNING: Unable to find provider 'fusilli-mate-gtk' of required component 'windowmanager'
```

and for firefox
```
(firefox:28576): GLib-GObject-WARNING **: Attempt to add property GnomeProgram::sm-connect after class was initialised

(firefox:28576): GLib-GObject-WARNING **: Attempt to add property GnomeProgram::show-crash-dialog after class was initialised

(firefox:28576): GLib-GObject-WARNING **: Attempt to add property GnomeProgram::display after class was initialised

(firefox:28576): GLib-GObject-WARNING **: Attempt to add property GnomeProgram::default-icon after class was initialised

(firefox:28576): Gtk-CRITICAL **: _gtk_text_layout_get_block_cursor: assertion 'layout != NULL' failed

(firefox:28576): Gtk-CRITICAL **: IA__gtk_text_layout_get_cursor_locations: assertion 'layout != NULL' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gtk-CRITICAL **: _gtk_text_layout_get_block_cursor: assertion 'layout != NULL' failed

(firefox:28576): Gtk-CRITICAL **: IA__gtk_text_layout_get_cursor_locations: assertion 'layout != NULL' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gtk-CRITICAL **: _gtk_text_layout_get_block_cursor: assertion 'layout != NULL' failed

(firefox:28576): Gtk-CRITICAL **: IA__gtk_text_layout_get_cursor_locations: assertion 'layout != NULL' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gtk-CRITICAL **: _gtk_text_layout_get_block_cursor: assertion 'layout != NULL' failed

(firefox:28576): Gtk-CRITICAL **: IA__gtk_text_layout_get_cursor_locations: assertion 'layout != NULL' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gtk-CRITICAL **: _gtk_text_layout_get_block_cursor: assertion 'layout != NULL' failed

(firefox:28576): Gtk-CRITICAL **: IA__gtk_text_layout_get_cursor_locations: assertion 'layout != NULL' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gtk-CRITICAL **: _gtk_text_layout_get_block_cursor: assertion 'layout != NULL' failed

(firefox:28576): Gtk-CRITICAL **: IA__gtk_text_layout_get_cursor_locations: assertion 'layout != NULL' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gtk-CRITICAL **: _gtk_text_layout_get_block_cursor: assertion 'layout != NULL' failed

(firefox:28576): Gtk-CRITICAL **: IA__gtk_text_layout_get_cursor_locations: assertion 'layout != NULL' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gtk-CRITICAL **: _gtk_text_layout_get_block_cursor: assertion 'layout != NULL' failed

(firefox:28576): Gtk-CRITICAL **: IA__gtk_text_layout_get_cursor_locations: assertion 'layout != NULL' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gtk-CRITICAL **: _gtk_text_layout_get_block_cursor: assertion 'layout != NULL' failed

(firefox:28576): Gtk-CRITICAL **: IA__gtk_text_layout_get_cursor_locations: assertion 'layout != NULL' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gtk-CRITICAL **: _gtk_text_layout_get_block_cursor: assertion 'layout != NULL' failed

(firefox:28576): Gtk-CRITICAL **: IA__gtk_text_layout_get_cursor_locations: assertion 'layout != NULL' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed

(firefox:28576): Gdk-CRITICAL **: gdk_window_invalidate_rect_full: assertion 'GDK_IS_WINDOW (window)' failed
```

**Comment by noodlylight** [comment]
> Ok, i found a hint in .x-session-errors .
> 
> mate-session[27944]: WARNING: Unable to find provider 'fusilli-mate-gtk' of required component 'windowmanager'

I think this means that the script (fusilli-mate-gtk) is not started at all.

I have no idea about firefox. I suspect there is some kind of hardware acceleration conflict. Can you check what happens if you disable it ? (http://lifehacker.com/disable-firefoxs-hardware-acceleration-to-fix-slowness-749344037)

**Comment by NiceandGently** [comment]
Disable hardware acceleration in firefox didn't help, it is impossible to to type this text here.
But i noticed also probs with other application. Here the mouse hover function in menus or list views has a delay or stays on an entry if i already on another entry. I.e dconf-editor is unusable, i can't browes through the tree there.
Anyways i need to test it on another system.

@ script
Yes, mate-session-manger don't find or accept the script for some reason.
Couriously that it works with the compiz-mate-gtk script which is basicly the same.

**Comment by noodlylight** [comment]
> Disable hardware acceleration in firefox didn't help, it is impossible to type this text here.
> But i noticed also probs with other application. Here the mouse hover function in menus or list views has a delay or stays on an entry if i'm already on another entry with the mouse. I.e dconf-editor is unusable, i can't browes through the tree there.
> Anyways, i need to test it on another system.

You can also check with a new, blank banana.xml (but keep the old one for comparison). It could be an option, or one of the plugins I recently ported.

> @ script
> Yes, mate-session-manger don't find or accept the script for some reason.
> Couriously that it works with the compiz-mate-gtk script which is basicly the same.

Permissions?

**Comment by NiceandGently** [comment]
Starting with a fresh configuration doesn't help, i can't write this post and i've rendering probs with applications, ie, fsm !
![bildschirmfoto-fusilli settings manager --- mate compatibility](https://cloud.githubusercontent.com/assets/961604/6651771/5d763678-ca54-11e4-8adf-7c168e5a83dc.png)
You see the weird buttons, also switching between tabs inside fsm don't work very well, i need to do several clicks to switch to another tab, and i don't see every entries in some tabs. I fi click on a visible entry than the other will display.
It's a bit hard to discribe what's going on here.
I noticed that fusilli starts with cairo decorator with the blank conf. Inside fsm i needed to disable 'auto-save' for that fsm could write in bannana.xml and use gtk-decorator.
In the new configuration i enabled only the wall plugin but the render issues were there before i enabled the plugin.
As i said before i need to run fusilli on another system for more testing, but here on my productive system it's impossible to use it currently.
Maybe the workaround plugin is needed from compiz?

PS: i use proprietary nvidia driver.

**Comment by noodlylight** [comment]
I just tested fusilli on a nvidia system (using the proprietary driver).

There are quite a lot of artifacts, but I can type fine. 

I also tried compiz-0.8.8 and I get the same artifacts, even with the workaround plugin enabled (force sync between X and GLX).

**Comment by NiceandGently** [comment]
Change topic title, session start script is minor.
My compiz-0.88 version is heavily patched for fedora as you know.

**Comment by noodlylight** [comment]
Good! 

I'm willing to try fedora and the patched compiz when I find the time. Where should I look? Should I just install fedora and compiz from it's repositories?

**Comment by NiceandGently** [comment]
Easiest way to install fedora with Mate + Compiz is to use a spin.
http://spins.fedoraproject.org/mate-compiz/
In the menu you'll find entries for starting compiz.

**Comment by noodlylight** [comment]
Installed Fedora 21 MATE Compiz Spin + NVIDIA 346.47.

Compiz works fine.

Fusilli also works fine. 

This is the same system where Gentoo + NVIDIA leads to horrible artifacts on both compiz and fusilli.

Bananas.

**Comment by noodlylight** [comment]
I just fixed compiz/fusilli on my Gentoo install too... 

There was an issue with the nvidia drivers, reinstalling them fixed the artifacts.

So it seems I cannot reproduce this bug. :(

**Comment by NiceandGently** [comment]
Thanks for trying to reproduce it.
Today i installed fusilli in another fresh fedora installation (next release, f22 alpha) on the same system.
I saw the same issue but i could finally track down what the issue is :)
The screen won't be updated if i do a mouse action in an application, ie. switching to another tab. If i move the window a bit than i see the content of the new  tab.
A little video which demonstrate the issue with fsm and caja preferences window.
https://dl.dropboxusercontent.com/u/49862637/Mate-desktop/Bugs/fusilli.ogv
I'm not shure if this is cause by legacy nvidia-340.xxx driver, which i need to use since Jan/Feb because my GT240 isn't supported anymore with latest driver.
It looks more that a setting is missing or whatever.
My fusilli conf.
https://dl.dropboxusercontent.com/u/49862637/Mate-desktop/Bugs/banana.xml
I will try fusilli on my notebook and intel driver.

**Comment by noodlylight** [comment]
Thanks a lot for investigating and recording a video of the bug!

Unfortunately I don't know where to look. I haven't changed anything in the critical compositing/OpenGL parts, so it seems weird to me that it happens on fusilli and not compiz. Maybe plugin workaround has something related.

**Comment by NiceandGently** [comment]
Good news, on my notebook with intel driver fusilli runs well :)
So it looks like a prob with nvidia-340.xxx legacy driver, maybe.
I will downgrade nvidia to nvidia-304.xxx legacy driver the next days to confirm my theory.
If i remember correctly it wasn't a prob with that driver which i used before nvidia-340 was available at rpmfusion in january.

**Comment by NiceandGently** [comment]
The issue is gone with nvidia-3.04.xxx legacy driver :)
So, maybe the culprit is nvidia-340.xxx legacy driver itself, or there is a package packing issue with the rpmfusion version, or the workaround plugin is needed.
I will test nvidia-340.xxx directly from nvidia site the next days, in order to exclude a packaging issue with rpmfusion driver.

**Comment by NiceandGently** [comment]
I could also solved the issue with using my start script for session start. A simply desktop file is needed in /usr/share/applications.  :)
If you like this idea i can do a pull request for the script + desktop file.

**Comment by noodlylight** [comment]
> The issue is gone with nvidia-3.04.xxx legacy driver :)
> So, maybe the culprit is nvidia-340.xxx legacy driver itself, or there is a packaging issue with the rpmfusion version, or the workaround plugin is needed.
> I will test nvidia-340.xxx directly from nvidia site the next days, in order to exclude a packaging issue with rpmfusion driver.

Good, thanks a lot.



> I could also solved the issue with using my start script for session start. A simply desktop file is needed in /usr/share/applications. :)
> If you like this idea i can do a pull request for the script + desktop file.

Sure, sounds a good idea.

**Comment by noodlylight** [comment]
Just found this: https://bugs.freedesktop.org/show_bug.cgi?id=27420

Working on porting workarounds plugin right now.

**Comment by noodlylight** [comment]
Added plugin workarounds at a336e1f07581e0940891b076c49aecdfb3d7b53d

**Comment by NiceandGently** [comment]
Congratulations,
with https://github.com/noodlylight/fusilli/commit/7bc6e6e56adb51c687af28704f21db1fb48ca4e2
and the workaround plugin the issue is gone :)
But i didn't enable 'Fix screen updates in XGL with fglrx' which was mentioned in the freedesktop report, only the default workarounds.
But i will test more with nvidia-340.xxx directly from nvidia site.
I can't imagine that 340 has a regression which current drivers or 304.xxx don't have.
Anyways, the issue is fixed.

Thank you

**Comment by NiceandGently** [comment]
Ok, the issue exists also with orign nvidia driver, but not so extrem as with driver from rpmfusion.

**Comment by noodlylight** [comment]
> Congratulations,
> with 7bc6e6e
> and the workaround plugin the issue is gone :)
> But i didn't enable 'Fix screen updates in XGL with fglrx' which was mentioned in the freedesktop report, only the default workarounds.
> But i will test more with nvidia-340.xxx directly from nvidia site.
> I can't imagine that 340.xxx has a regression which current drivers or 304.xxx don't have.
> Anyways, the issue is fixed.
> 
> Thank you

Sweet!

Did you find which option exactly fixes the bug ? (i mean the options in workarounds)

**Comment by NiceandGently** [comment]
> Did you find which option exactly fixes the bug ? (i mean the options in workarounds)

It's 'Force syncronisation between X and GLX'

**Comment by NiceandGently** [comment]
With that fix i don't need to enable workarounds plugin.
I use this fix also for compiz-0.88 in fedora.
